src/CopyOnWrite/Copy.cs

```diff
@@ -976,25 +1009,64 @@ private static int ParseIntFromEnvironmentVariableOrDefault(string environmentVa
                 : defaultValue;
         }
 
+        private void LogWarning(string message, params object[] messageArgs)
+        {
+            string code = ExtractMessageCode(message);
+            Log.LogWarning(subcategory: null, warningCode: code, helpKeyword: null, file: null, lineNumber: 0, columnNumber: 0, endLineNumber: 0, endColumnNumber: 0, helpLink: null, message: message, messageArgs: messageArgs);
+        }
+
+        private void LogError(string message, params object[] messageArgs)
+        {
+            string code = ExtractMessageCode(message);
+            Log.LogError(subcategory: null, errorCode: code, helpKeyword: null, file: null, lineNumber: 0, columnNumber: 0, endLineNumber: 0, endColumnNumber: 0, helpLink: null, message: message, messageArgs: messageArgs);
+        }
+
+        private static string ExtractMessageCode(string message)
```

Comment:
MSBuild's TaskLoggingHelper has a public API for this, can we use it?
https://github.com/dotnet/msbuild/blob/08494c73128451a3f7cfb47a5e9cbd63f5507a1f/src/Shared/TaskLoggingHelper.cs#L178

Reply:
It appears that the code wants to pull the resources from the source assembly, not the core MSBuild assembly, which is why @AndyGerlicher originally broke away from using that API.